clients/src/main/java/org/apache/kafka/common/protocol/Errors.java

```diff
@@ -192,10 +192,18 @@ public static Errors forCode(short code) {
     }
 
     /**
-     * Return the error instance associated with this exception (or UNKNOWN if there is none)
+     * Return the error instance associated with this exception or any of its superclasses (or UNKNOWN if there is none).
+     * If there are multiple matches in the class hierarchy, the first match starting from the bottom is used.
      */
     public static Errors forException(Throwable t) {
-        Errors error = classToError.get(t.getClass());
-        return error == null ? UNKNOWN : error;
+        Errors error = null;
+        Class clazz = t.getClass();
+        while (clazz != null) {
+            error = classToError.get(clazz);
```

Comment:
Tiny nitpick: you can declare error in this line too as it's not used outside the loop anymore